backend/app/routes/web/research_product.py

```diff
@@ -0,0 +1,57 @@
+"""Endpoint for adding a research product to favourites"""
+import asyncio
+import logging
+from contextlib import suppress
+from typing import List, Optional
+
+from fastapi import APIRouter, Depends
+from httpx import AsyncClient, ConnectError
+from pydantic import AnyUrl
+
+from app.schemas.research_product_response import Collection, ResearchProductResponse
+from app.solr.operations import get_dep
+
+router = APIRouter()
+
+logger = logging.getLogger(__name__)
+
+
+@router.get("/research-product/{resource_type}/{rp_id}")
+async def get_rp_by_id(
+    resource_type: Collection, rp_id: str, solr_get=Depends(get_dep)
+) -> ResearchProductResponse:
+    """
+    Main function responsible for getting details for a given Solr document.
+    args:
+        resource_type: string literal - one of permitted collections
+        rp_id: ID of a research product to be found
+    """
+    async with AsyncClient() as async_client:
+        response = await solr_get(async_client, resource_type, rp_id)
+        response = response.json()["doc"]
+
+    return ResearchProductResponse(
+        title=" ".join(response["title"]),
+        links=await _validate_urls(response["url"]),
+        author=response["author_names"],
+        type=response["type"],
+    )
+
+
+async def _validate_urls(urls: List[AnyUrl]) -> List[Optional[AnyUrl]]:
+    """Runs a validation for every url for the product and returns
+    only the valid ones.
+    """
+    valid_urls = await asyncio.gather(*[_validate_url(url) for url in urls])
+    return list(filter(None, valid_urls))
+
+
+async def _validate_url(url: AnyUrl) -> Optional[AnyUrl]:
+    """Performs a request to check a link's validity"""
+    response = None
+    async with AsyncClient() as async_client:
+        with suppress(ConnectError):
+            response = await async_client.get(url, follow_redirects=True)
```

Comment:
This call might rise exceptions - https://www.python-httpx.org/exceptions/
We should catch any exception and treat it as an error status.

Reply:
Network error is not the only exception. But either way - exception here will probably destroy "gather" call.
Additionally server type errors (malformatted https response) is also an error.
If it has been decided (I forgot) that this kind of issue should not be filtered out than we should catch that exception and return url (not `None`).

Current implementation will not return url which triggers error and potentially won't return any response in API.